packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,176 @@
+
+export default createRule({
+  name: 'thenable-in-promise-aggregators',
+  meta: {
+    docs: {
+      description:
+        'Disallow passing non-Thenable values to promise aggregators',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      inArray:
+        'Unexpected non-Thenable value in array passed to promise aggregator.',
+      arrayArg:
+        'Unexpected array of non-Thenable values passed to promise aggregator.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    const aggregateFunctionNames = ['all', 'race', 'allSettled', 'any'];
+
+    function skipChainExpression<T extends TSESTree.Node>(
+      node: T,
+    ): T | TSESTree.ChainElement {
+      return node.type === AST_NODE_TYPES.ChainExpression
+        ? node.expression
+        : node;
+    }
+
+    function isPartiallyLikeType(
+      type: ts.Type,
+      predicate: (type: ts.Type) => boolean,
+    ): boolean {
+      if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+        return true;
+      }
+      if (type.isIntersection() || type.isUnion()) {
+        return type.types.some(t => isPartiallyLikeType(t, predicate));
+      }
+      return predicate(type);
+    }
+
+    function isIndexableWithSomeElementsLike(
+      type: ts.Type,
+      predicate: (type: ts.Type) => boolean,
+    ): boolean {
+      if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+        return true;
+      }
+
+      if (type.isIntersection() || type.isUnion()) {
+        return type.types.some(t =>
+          isIndexableWithSomeElementsLike(t, predicate),
+        );
+      }
+
+      if (!checker.isArrayType(type) && !checker.isTupleType(type)) {
+        const indexType = checker.getIndexTypeOfType(type, ts.IndexKind.Number);
+        if (indexType === undefined) {
+          return false;
+        }
+
+        return isPartiallyLikeType(indexType, predicate);
+      }
+
+      const typeArgs = type.typeArguments;
+      if (typeArgs === undefined) {
+        throw new Error(
+          'Expected to find type arguments for an array or tuple.',
+        );
+      }
+
+      return typeArgs.some(t => isPartiallyLikeType(t, predicate));
+    }
+
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        const callee = skipChainExpression(node.callee);
+        if (callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        if (callee.computed) {
+          if (
+            callee.property.type !== AST_NODE_TYPES.Literal ||
+            typeof callee.property.value !== 'string' ||
+            !aggregateFunctionNames.includes(callee.property.value)
+          ) {
+            return;
+          }
+        } else if (!aggregateFunctionNames.includes(callee.property.name)) {
```

Comment:
[Refactor] This _"get the name one way if it's computed or another if it's static"_ need comes up a lot in rules. We should have a shared utility for it. `getStaticStringValue` should do the trick, or `getStaticValue`. One of those 😄.

Reply:
There is already a function `getStaticStringValue` in `eslint-plugin/src/util` but it's about finding what the source representation of a value would be, e.g., turning a regex into `/this/` or null into the string `null`, etc.

To avoid confusion with that I've broken out a `getMemberName` function you can look at and see if it might be worth breaking out into a shared utility